metapool/tests/test_sample_sheet.py

```diff
@@ -1732,6 +1742,456 @@ def test_metagenomic_sheet_creation(self):
+            sheet.add_sample(sample_sheet.Sample(dict(zip(header, row))))
+
+        # Once sheet has been manually populated, validate it.
+        self.assertTrue(sheet.validate_and_scrub_sample_sheet())
+        self.assertFalse(sheet.contains_katharoseq_samples())
+
+    def test_katharoseq_enabled_sheet_creation(self):
+        # create a sheet from scratch, this time with karathoseq samples.
+        sheet = MetagenomicSampleSheetv101()
+        sheet.Header['IEMFileVersion'] = 4
+        sheet.Header['sheetType'] = 'standard_metag'
+        sheet.Header['sheetVersion'] = '101'
+        sheet.Header['Investigator Name'] = 'Knight'
+        sheet.Header['Experiment Name'] = 'RKO_experiment'
+        sheet.Header['Date'] = '2021-08-17'
+        sheet.Header['Workflow'] = 'GenerateFASTQ'
+        sheet.Header['Application'] = 'FASTQ Only'
+        sheet.Header['Assay'] = 'Metagenomic'
+        sheet.Header['Description'] = ''
+        sheet.Header['Chemistry'] = 'Default'
+        sheet.Reads = [151, 151]
+        sheet.Settings['ReverseComplement'] = 0
+
+        data = [
+            ['Project1_99999', '99999', 'False', 'AACC', 'GGTT', 'False',
+             'False', 'protocol_1', 'a designed experiment']
+        ]
+
+        sheet.Bioinformatics = pd.DataFrame(
+            columns=['Sample_Project', 'QiitaID', 'BarcodesAreRC',
+                     'ForwardAdapter', 'ReverseAdapter', 'HumanFiltering',
+                     'contains_replicates', 'library_construction_protocol',
+                     'experiment_design_description'], data=data)
+
+        sheet.Contact = pd.DataFrame(columns=['Email', 'Sample_Project'],
+                                     data=[['c2cowart@ucsd.edu',
+                                            'Project1_99999'], ])
+
+        header = ['Sample_ID', 'Sample_Name', 'Sample_Plate', 'well_id_384',
+                  'I7_Index_ID', 'index', 'I5_Index_ID', 'index2',
+                  'Sample_Project', 'Well_description']
+
+        data = [
+            ['sample_1', 'sample.1', 'sample_plate_1', 'A1', 'iTru7_107_07',
+             'CCGACTAT', 'iTru5_01_A', 'ACCGACAA', 'Project1_99999', 'desc'],
+            ['sample_2', 'sample.2', 'sample_plate_1', 'A2', 'iTru7_107_07',
+             'CCGACTAC', 'iTru5_01_A', 'ACCGACAT', 'Project1_99999', 'desc'],
+            ['sample_3', 'sample.3', 'sample_plate_1', 'A3', 'iTru7_107_07',
+             'CCGACTAG', 'iTru5_01_A', 'ACCGACAG', 'Project1_99999', 'desc'],
+            # added katharoseq control here.
+            ['kath0001', 'kath0001', 'sample_plate_1', 'A4', 'iTru7_107_07',
+             'CCGCCTAG', 'iTru5_01_A', 'ACCGTCAG', 'Project1_99999', 'desc']
+        ]
+
+        for row in data:
+            # For all children of Samplesheet() class, the first call to
+            # add_sample() determines the number, name, and ordering of
+            # columns in the [Data] section. Changing the columns in a
+            # subsequent call will raise an Error.
```

Comment:
These seem like the same for-loop/comment, that lines 2075 onward; would it be easier to just make a copy of the object before the changes/test to minimize code?

Reply:
It's important to understand that the purpose of the tests is to confirm the behavior when creating a fresh, new object, first with the base set of columns and second with the base + optional columns. That being said, a new instance of KarathoseqEnabledSheetCreationTests() is being created each time a test is run so it's okay to partially initialize a SampleSheet() and make it a member of the class. I will do that now ty.